### ISSUE TITLE
Pods sometimes have no endpoints on re-created nodes

### ISSUE THREAD

**Issue by dbaumgarten** [cilium/cilium#19319]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Hi,
I have a bunch of Kubernetes-Clusters running cilium. Because of internal security-policies I have to regularly change the ubuntu-images the nodes are running.
This happens by draining a done, deleting it's VM, recreating the VM and joining it back into the cluster (for all the nodes in a rolling fashion).
This works just fine, at least most of the time.

Now and then after re-creating a node, the (non hostNetwork) pods running on that new node do not have any network connectivity.

Cilium reports everything is fine. Cilium-health an that node reports working connections to all other nodes.

When looking more closely at this I found out, that the affected pods do not seem have any cilium-endpoints.

```
kubectl -n kube-system get po -owide | grep minion-2
cilium-8pktk                                        1/1     Running            0                95m    192.168.3.9    daniel-test-kube-minion-2   <none>           <none>
csi-cinder-nodeplugin-6l5cq                         3/3     Running            0                94m    192.168.3.9    daniel-test-kube-minion-2   <none>           <none>
fluentd-n4cnm                                       0/1     CrashLoopBackOff   22 (4m55s ago)   94m    10.192.4.80    daniel-test-kube-minion-2   <none>           <none>
kube-proxy-t9554                                    1/1     Running            0                95m    192.168.3.9    daniel-test-kube-minion-2
```
(The flientd-pod is crashlooping because of missing network-connectivity. All other pods there are in the hostNetwork and are not affected).

When listing the cilium endpoints on that node:
```
 kubectl -n kube-system exec -it cilium-8pktk cilium endpoint list
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
Defaulted container "cilium-agent" out of: cilium-agent, clean-cilium-state (init)
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                           IPv6   IPv4           STATUS
           ENFORCEMENT        ENFORCEMENT
927        Disabled           Disabled          4          reserved:health                                              10.192.4.156   ready
2936       Disabled           Disabled          1          k8s:node.kubernetes.io/instance-type=gitc.xlarge.g3                         ready
                                                           k8s:topology.cinder.csi.openstack.org/zone=az-c
                                                           k8s:topology.kubernetes.io/region=de-wob-2
                                                           k8s:topology.kubernetes.io/zone=az-c
                                                           reserved:host

```
There is no endpoint for the fluentd-pod. (When looking at a functioning node I can clearly see an endpoint for it's fluent-pod)

Fluentd is not the only application that is affected by this (but it is a DeamonSet and needs the network, which causes it to trip over this quite often).

When examining the logs of the cilium-pod on the affected node I get this:
```
level=info msg="Create endpoint request" addressing="&{10.192.4.80 423faadc-8b4a-410a-8684-272328e89303  }" containerID=56b464d44afa7800bddc0589ac55b78f92eacb10281d723bbb6e5095ca11f181 datapathConfiguration="<nil>" interface=lxc98e1bace9
101 k8sPodName=kube-system/fluentd-n4cnm labels="[]" subsys=daemon sync-build=true
level=info msg="New endpoint" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3389 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (blocking)" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3389 identityLabels="k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=fluentd,k8s:io.kubernetes.pod.namespace=kube-system,k8s:k8s-app=fluentd-logging,k8s:version=v1" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Reusing existing global key" key="k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system;k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=fluentd;k8s:io.kubernetes.pod.namespace=kube-system;k8s:k8s-app=fluentd-logging;k8s:version=v1;" subsys=allocator
level=info msg="Identity of endpoint changed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3389 identity=30488 identityLabels="k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=fluentd,k8s:io.kubernetes.pod.namespace=kube-system,k8s:k8s-app=fluentd-logging,k8s:version=v1" ipv4= ipv6= k8sPodName=/ oldIdentity="no identity" subsys=endpoint
level=info msg="Waiting for endpoint to be generated" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3389 identity=30488 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="New endpoint" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=927 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (blocking)" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=927 identityLabels="reserved:health" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Identity of endpoint changed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=927 identity=4 identityLabels="reserved:health" ipv4= ipv6= k8sPodName=/ oldIdentity="no identity" subsys=endpoint
level=info msg="Compiled new BPF template" BPFCompilationTime=1.238736156s file-path=/var/run/cilium/state/templates/5a5fb6a2d306e4c7bc2c20df2403b467f64147176f1c0bf66ccab804a50c16fb/bpf_host.o subsys=datapath-loader
level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=2936 identity=1 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Compiled new BPF template" BPFCompilationTime=3.487575559s file-path=/var/run/cilium/state/templates/e12467916c7732f0434c407bd31d1fef0c9b4f8ef913858ff119a33123138281/bpf_lxc.o subsys=datapath-loader
level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=927 identity=4 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3389 identity=30488 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Releasing key" key="[k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=fluentd k8s:io.kubernetes.pod.namespace=kube-system k8s:k8s-app=fluentd-logging k8s:version=v1]" subsys=allocator
level=info msg="Removed endpoint" containerID= datapathPolicyRevision=1 desiredPolicyRevision=1 endpointID=3389 identity=30488 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=warning msg="Creation of endpoint failed" containerID= datapathPolicyRevision=1 desiredPolicyRevision=1 endpointID=3389 error="timeout while waiting for initial endpoint generation to complete" identity=30488 ipv4= ipv6= k8sPodName=/ subsys=daemon
level=info msg="API call has been processed" error="timeout while waiting for initial endpoint generation to complete" name=endpoint-create processingDuration=3.642088473s subsys=rate totalDuration=3.642166757s uuid=a6d02389-c1df-4929-b942-c98560cb413b waitDurationTotal=0s
level=info msg="Waiting for Hubble server TLS certificate and key files to be created" subsys=hubble
level=info msg="Serving cilium health API at unix:///var/run/cilium/health.sock" subsys=health-server
level=info msg="New encryption keys reclaiming SPI" SPI=3 oldSPI=0 subsys=ipsec
level=info msg="Conntrack garbage collector interval recalculated" deleteRatio=0.0003662109375 newInterval=7m30s subsys=map-ct
level=warning msg="Released IP after expiration" ip=10.192.4.80 subsys=ipam uuid=423faadc-8b4a-410a-8684-272328e89303
```

It seems like cilium tries to create that endpoint, but fails with "timeout while waiting for initial endpoint generation to complete" and just deletes it again.

Pods that are newly created on the node work just fine. And deleting the broken pods (which leads to their re-creation) also results in working new pods.

Whether or not this happens seems to be random. It is rare enough to be a pain to debug, but common enough to be an issue.
The random-nature makes it hard to pinpoint when exactly this started, but it did not appear before we switched form cilium 1.9.x to 1.10.x .

After hours of debugging this without any results, I came to the conclusion that this might be a bug in cilium and I need help.
"Help me ~Obi Wan Kenobi~ Cilium-Team, you're my only hope"

### Cilium Version

Client: 1.10.9 0338ca7 2022-03-26T01:51:32+01:00 go version go1.16.15 linux/amd64
Daemon: 1.10.9 0338ca7 2022-03-26T01:51:32+01:00 go version go1.16.15 linux/amd64

### Kernel Version

5.4.0-105-generic #119-Ubuntu SMP Mon Mar 7 18:49:24 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.23.0
(But this also seems to appear on v1.21.*)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by pchaigno** [comment]
Thanks for the report! Could you upload a Cilium sysdump with a node where this happens? If possible, it would also help to enable debug mode to get more logs on what could be happening here.

**Comment by dbaumgarten** [comment]
Here is the sysdump: [cilium-sysdump-20220405-113518.zip](https://github.com/cilium/cilium/files/8417710/cilium-sysdump-20220405-113518.zip)
I will enable debugmode, try to reproduce this and provide the logs as soon as I get them. Because of the random nature of this it might take little time.

**Comment by pchaigno** [comment]
We're missing the reason for the timeout in the code. I've sent a PR to fix that: https://github.com/cilium/cilium/pull/19333.

Debug-mode logs might still be useful if you get them.

**Comment by dbaumgarten** [comment]
So, I enabled debug-logging, used a self-compiled version of cilium 1.10.9 which includes your PR and reproduced the issue.
This is the result: [debuglog.txt](https://github.com/cilium/cilium/files/8425814/debuglog.txt)

**Comment by pchaigno** [comment]
Thanks, that helps! Would you have the same logs but with timestamps?

**Comment by dbaumgarten** [comment]
Sure, here: [debuglog.txt](https://github.com/cilium/cilium/files/8426376/debuglog.txt)

**Comment by aanm** [comment]
Did you check fluentd previous container logs and kubelet logs for any clues? From what I can tell, endpoint creation is being canceled by kubelet itself.

**Comment by dbaumgarten** [comment]
This is the first time that pod is started on the brand new node. There isn't any previous container.
I had looked at the kubelet logs, but didn't see anything that caught my attention.
Looking at the kubelet-logs again, I see that the kubelet is beeing restarted by some script if mine, right in the middle of the initial boot-process.
Maybe, when kubelet is restarted just at the wrong time during container-creation it causes endpoint-creation to be aborted?
I will try to make sure that kubelet can run uninterrupted and will see if that stops the issues.

**Comment by dbaumgarten** [comment]
Ok, after making sure kubelet can run uninterrupted, I tried reproducing the issue.
I have tried quite a few time, but the issue never occurred again.
The random nature of it makes it hard to say if it's actually gone, but I am getting more and more confident.

Tl;dr: Do not restart kubelet while it is trying to create pods, or they might end up without endpoints.

**Comment by Paulius0112** [comment]
Hi. This is still an issue. How did you solve it?
